### PR TITLE
Add `protocol` property to BusABC to determine active CAN Protocol

### PR DESCRIPTION
**Pull request by lumagi**
Something that I have been missing in the past about the generic `BusABC` is to introspect CAN-FD support independent of the bus implementation. The way I would imagine this is to simply have an attribute called `is_fd` or something similar that will return `True` if the bus supports the CAN-FD protocol.

As an example, I have implemented this for the PCAN bus to gather some feedback. If this gets support, I will implement it for all interfaces.

### CODE REVIEW

**Comment by zariiii9003** [comment]
Just to clarify: you say

> return True if the bus supports the CAN-FD protocol

but you actually just want to know whether CAN FD is currently active, right?

Generally i don't mind standardizing this for all interfaces, but with CAN XL coming soon it might make sense to use an enum instead of a boolean flag. Otherwise we would need to add `is_xl` soon.

**Comment by zariiii9003** [comment]
CAN FD is already part of ISO 11898 and CAN XL will also be part of 11898. So instead of `CAN_ISO_11898` i'd suggest `CAN_20` or something like that. Do you have any ideas for the class name? `CanSpec`? `Protocol`? `PhysicalLayer`? 

@hardbyte not sure how to contact you but i recommend @lumagi as maintainer ;)

**Comment by hardbyte** [comment]
@lumagi - Lukas thanks so much for your contributions so far. As suggested I've invited you to the repository as a collaborator. Use your new powers for good! In essence, please be sensible and request reviews for non-trivial changes 👍

Great idea @zariiii9003!

**Comment by lumagi** [comment]
@hardbyte Thank you for the invitation, I appreciate it! And thanks for the recommendation @zariiii9003  ;)

**Comment by lumagi** [comment]
Regarding the naming of the enum, I think I like `Protocol`. But I would like make it a bit more specific like `CANProtocol`, otherwise this might clash with a different symbol when importing it into the module scope.

**Comment by zariiii9003** [comment]
[Here](https://www.can-cia.org/can-knowledge/can/can-xl/) is an interesting sentence

> The third generation of the CAN data link layer supports all three protocol types (Classical CAN, CAN FD, and CAN XL)

So it is either "DataLinkLayer" or "Protocol". I'm not sure which one is technically correct in our case. I prefer `CanProtocol` over `BusProtocol` btw if we take the protocol option.

**Comment by lumagi** [comment]
Ok, to me either one would be fine: `CANLinkLayer`, `CANProtocol`. But I wouldn't make the names any longer.

File: can/interfaces/udp_multicast/bus.py
Comment (lumagi) [inline can/interfaces/udp_multicast/bus.py:109]: @zariiii9003 Would you consider this `is_fd` flag as a public attribute that would lead to a breaking API change if it were removed? With the `protocol` field it's technically redundant and could be removed.

Comment (zariiii9003) [reply can/interfaces/udp_multicast/bus.py:109]: I guess we could replace it with a property and raise a DeprecationWarning with a removal scheduled for version 5.

**Comment by zariiii9003** [comment]
> Ok, to me either one would be fine: `CANLinkLayer`, `CANProtocol`. But I wouldn't want to make the names too long.

I just looked through the rest of the code, can you rename it to `CanProtocol`? That's more in line with the rest of our code:
image

**Comment by lumagi** [comment]
Done

**Comment by zariiii9003** [comment]
That's a big one. We'll probably have to release 4.2 and then leave this one in develop for some time 😄

**Comment by lumagi** [comment]
Yes sure, I don't mind. I only have a couple Bus implementations left, so this can probably change state away from draft soon. But I imagine there's gonna be some feedback :D

**Comment by zariiii9003** [comment]
I resolved all the merge conflicts from #1551. I hope i didn't break anything.

**Comment by lumagi** [comment]
Thanks for resolving the conflicts! And also the rogue black test - It was not able to come to terms with my local black version.

**Comment by zariiii9003** [comment]
> Thanks for resolving the conflicts! And also the rogue black test - It was not able to come to terms with my local black version.

You could install the correct versions with `pip install -r requirements-lint.txt`

**Comment by lumagi** [comment]
@zariiii9003 I think I'm through with all the implementations, if you would like to take a look.

File: can/bus.py
Comment (zariiii9003) [inline can/bus.py:60]: I'm not sure about this parameter here. I think subclasses should keep the parameters of its superclasses, which wouldn't make sense in this case.

What do you think about declaring `protocol` as a class attribute?

Comment (lumagi) [reply can/bus.py:60]: Maybe as a protected instance attribute, i.e. (self._protocol)? If it was a class attribute, then it would have the same value for all instances, right?

Comment (zariiii9003) [reply can/bus.py:60]: We could set the default value of `_protocol` as a class attribute. In the `__init__` method, the new value is assigned to the instance, not the class. So multiple instances will not share the same value.

Comment (lumagi) [reply can/bus.py:60]: Not passing it in the super constructor but setting it directly sounds like a good plan. I assume you want to set it as a class attribute as a fallback in case it is not set in the subclass constructor? How would you be able to set it as an instance attribute if it is already present as a class attribute?

Comment (zariiii9003) [reply can/bus.py:60]:
```python3
class A:
    protocol = "Class Attribute"


class B(A):
    def __init__(self):
        self.protocol = "Instance Attribute"
        super().__init__()


print(A().protocol)
print(B().protocol)
```

This prints 
```text
Class Attribute
Instance Attribute
```

Comment (lumagi) [reply can/bus.py:60]: Great, I wasn't aware that the way the attribute was accessed for assignment changed the assignment scope. I will change the implementation to follow this approach.

Comment (zariiii9003) [reply can/bus.py:60]: @lumagi tell me if you need help with the changes

Comment (lumagi) [reply can/bus.py:60]: Thanks for the offer. I have been a bit preoccupied with my time the last weeks, but I should be able to make some progress this week.

Comment (lumagi) [reply can/bus.py]: The attribute should now be set inside the subclass constructor for all classes.

File: can/interfaces/pcan/pcan.py
Comment (zariiii9003) [inline can/interfaces/pcan/pcan.py:493]:
```suggestion
            if self._can_protocol is CanProtocol.CAN_FD:
```
properties are rather slow, and enums are compared with `is`

File: can/interfaces/pcan/pcan.py
Comment (zariiii9003) [inline can/interfaces/pcan/pcan.py:601]:
```suggestion
        if self._can_protocol is CanProtocol.CAN_FD:
```

File: can/interfaces/udp_multicast/bus.py
Comment (zariiii9003) [inline can/interfaces/udp_multicast/bus.py:139]:
```suggestion
        if self._can_protocol is not CanProtocol.CAN_FD and can_message.is_fd:
```

File: can/interfaces/udp_multicast/bus.py
Comment (zariiii9003) [inline can/interfaces/udp_multicast/bus.py:145]:
```suggestion
        if self._can_protocol is not CanProtocol.CAN_FD: and msg.is_fd:
```

File: can/interfaces/vector/canlib.py
Comment (zariiii9003) [inline can/interfaces/vector/canlib.py:666]:
```suggestion
                if self._can_protocol is CanProtocol.CAN_FD:
```

File: can/interfaces/vector/canlib.py
Comment (zariiii9003) [inline can/interfaces/vector/canlib.py:800]:
```suggestion
        if self._can_protocol is CanProtocol.CAN_FD:
```

**Review by zariiii9003 — APPROVED** [review]
Looks great! The enum comparisons should be fixed though 😄

**Comment by lumagi** [comment]
Great, thanks for the review! How many reviewers do you normally have for greater changes like this one?

**Comment by zariiii9003** [comment]
> Great, thanks for the review! How many reviewers do you normally have for greater changes like this one?

At some point 2 reviews were required. But we had more active maintainers at that point. I will merge this after a 4.2.1 bugfix release to avoid backporting headaches.

**Comment by lumagi** [comment]
@zariiii9003 When do you think this will make it into a release? Will this have to wait until a major release or will this go into the next minor? To be honest, I am quite looking forward to having this feature available :D

**Comment by zariiii9003** [comment]
@lumagi I was hoping to include #1611. But i could release 4.3 anytime. I'll try to get it ready by the end of the week.
You could accelerate it by updating the changelog to include all relevant changes since 4.2.2 😄

**Comment by lumagi** [comment]
Sure, I can take care of that. I'll probably get around to doing that over the weekend.

**Pull request merged**
